### PR TITLE
add accessibility help extension contributions pt

### PR DESCRIPTION
**Pull request by meganrogge**
This includes:

- an accessibility help dialog implementation for each view with `accessibleHelpContent`
- an accessibility help hint that is applied to the view's aria label until a user opens the accessibility help dialog for that view. this required adding a new `AcessibleViewInformationService` to `workbench/services` so the view service can understand if an accessible view has been shown already in that layer
- resolves `<keybinding:commandId>` to the keybinding, if any, or a message and linked command to configure the keybinding

Screenshot 2024-04-15 at 2 09 28 PM

### CODE REVIEW

File: src/vs/workbench/contrib/accessibility/browser/accessibleView.ts
Comment (alexr00) [inline src/vs/workbench/contrib/accessibility/browser/accessibleView.ts:358]: I see the `ACCESSIBLE_VIEW_SHOWN_` prefix used in several places. It would be nice to extract this into a constant.

File: src/vs/workbench/contrib/accessibility/browser/accessibleView.ts
Comment (alexr00) [inline src/vs/workbench/contrib/accessibility/browser/accessibleView.ts:363]: This is just personal preference, but I prefer having an `isIAccessibleContentProvider` function over checking for the existence of a property.

File: src/vs/workbench/browser/parts/views/viewPane.ts
Comment (alexr00) [inline src/vs/workbench/browser/parts/views/viewPane.ts:554]: Having the `ViewPane` check the storage service to learn whether the accessible view is showing feels like implementation of accessible view leaking into here. Can the the accessible view service have an `isShowing` method?

File: src/vs/workbench/contrib/accessibility/browser/accessibleView.ts
Comment (alexr00) [inline src/vs/workbench/contrib/accessibility/browser/accessibleView.ts:510]: Having an `isXYZ` method here instead of `'next' in provider` would mean that you don't need to check for `'previous'` too.

**Review by alexr00 — DISMISSED** [review]
Very cool! In addtion to my inline comments, why is the `accessibilityHelpContent` contribution only a string and not also a markdown string?

File: src/vs/workbench/contrib/accessibility/browser/accessibleView.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/accessibility/browser/accessibleView.ts:363]: I agree, will do 👍🏼

File: src/vs/workbench/browser/parts/views/viewPane.ts
Comment (meganrogge) [reply src/vs/workbench/browser/parts/views/viewPane.ts:554]: View pane cannot access the layer where the accessible view is. I tried moving the accessible view and it's not possible AFAICT

Comment (meganrogge) [reply src/vs/workbench/browser/parts/views/viewPane.ts:554]: I could set a context key though instead of using the storage service

**Comment by meganrogge** [comment]
> why is the accessibilityHelpContent contribution only a string and not also a markdown string?

I was making it into one eventually, but agree, we should allow it to be at the extension contribution point.

File: src/vs/workbench/browser/parts/views/viewPane.ts
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/viewPane.ts:555]: Could we have an `AccessibleViewService` that lives in `src\vs\workbench\services`?

File: src/vs/workbench/contrib/files/browser/views/openEditorsView.ts
Comment (meganrogge) [inline src/vs/workbench/contrib/files/browser/views/openEditorsView.ts:97]:
```suggestion
		@IOpenerService openerService: IOpenerService,
```

File: src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts
Comment (mjbvz) [inline src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts:408]: Instead of inventing a new syntax, can we re-use the markdown link syntax? This could either be a normal markdown link:

```
[optional extra text](command:commandId....)

or maybe:

[optional extra text](keybinding:commandId)

```

Or use the auto link syntax:

```
<command:commandId....>

<vscode-keybinding:commandId....>
```

The advantage is that we get the parsing essentially for free and it will be much more reliable than string replacements

**Review by mjbvz — COMMENTED** [review]
Let's discuss how keybindings should be expressed at an API sync before merging this

File: src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts:408]: Yes, makes sense. Far preferable if we don't have to do extra parsing. I was not aware of the auto link syntax.

Comment (meganrogge) [reply src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts]: I also wonder if we should have syntax which allows someone to just provide the command id and have it render [Command Title] (keybinding).

When I tested contributing this to a view, it got redundant quickly:
```
"accessibilityHelpContent": "Helpful commands include: \n - Refresh Issue List (kb(command:issue.refresh))\n-GitHub Pull Requests: Focus on Issues View (kb(command:issues:github.focus)) \n-GitHub Pull Requests: Focus on Pull Requests View (kb(command:pr:github.focus))\n-GitHub Issues: Copy GitHub Permalink (kb(command:issue.copyGithubPermalink))\n-GitHub Issues: Create an Issue (kb(command:issue.createIssue))\n-GitHub Pull Requests: Create Pull Request (kb(command:pr.create))\n-GitHub Pull Requests: Start Review (kb(command:pr.startReview))\n-GitHub Pull Requests: Mark Pull Request Ready For Review (kb(command:pr.readyForReview))\n-GitHub Pull Requests: Approve (kb(command:review.approve))"
```

Comment (meganrogge) [reply src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts]: I'm not sure where the auto-link syntax is used 🤔

Comment (meganrogge) [reply src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts]: I am changing it to `(keybinding:commandId)`.

Comment (meganrogge) [reply src/vs/workbench/contrib/accessibility/browser/accessibleViewContributions.ts]: I have also added support for configuring the keybinding when there isn't one.

**Comment by meganrogge** [comment]
To discuss in API sync:
1. Do we want to have syntax which allows just a command ID to be provided and resolved as a the command title, its keybinding (if any) or the link to configure the keybinding
2. Can we provide a link for just text without the need for `[command name](command id)` syntax so screen reader users don't have to navigate through all of that? same question, but for the keybinding

File: src/vs/workbench/browser/parts/views/viewPane.ts
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/viewPane.ts:554]: I was thinking more that the entirety of this class could move into `vs\workbench\services`, but I'm not going to block the PR on that:

https://github.com/microsoft/vscode/blob/0ba604cc9d59e63419f3dadf0a29b683ccf55c08/src/vs/workbench/contrib/accessibility/browser/accessibleView.ts#L832-L833

**Review by alexr00 — DISMISSED** [review]
You have my approval, but I unchecked auto-merge since Matt wants to discuss at the API sync.

**Comment by meganrogge** [comment]
@mjbvz discussed with Jo at the API sync and he has signed off on what I'm doing for keybindings.

I will get feedback on and defer this part:

> Can we render the markdown such that a screen reader user doesn't have to navigate through [Configure a Keybinding](commandLink)?

**Review by rzhao271 — APPROVED** [review]

**Pull request merged**
